server/etcdserver/txn/txn_test.go

```diff
@@ -367,6 +370,26 @@ func TestWriteTxnPanic(t *testing.T) {
 		},
 	}
 
+	// compute DB file hash before applying the txn
+	dbHashBefore, err := computeFileHash(bePath)
+	if err != nil {
+		t.Fatalf("failed to compute DB file hash before txn: %v", err)
+	}
+
+	// we verify the following things below:
+	// - server panics after a write txn aply fails (invariant being tested: server should never try to move on from a failed write)
+	// - no writes from the txn are applied to the backend (invariant being tested: failed write should have no side-effect on DB state besides panic)
+	defer func() {
```

Comment:
Don't think there is a need to defer this? `assert.Panics` should handle panic?

Reply:
Did not follow @serathius 's comment above. 

- The purpose here is to verify the hash of the db did not change after the panicking. It's good!
- When panicking happens, any functions [deferred](https://go.dev/ref/spec#Defer_statements) are then executed as usual

Reply:
that's correct, test function stops executing after assert.panic - so further verification needs to happen either inside a defer or a t.Cleanup()

Reply:
I think what @serathius said is that the `assert.Panics` is capable to recover the panic. there is no need to use defer.
We should check the hash after `assert.Panics`.

Reply:
FYI https://github.com/stretchr/testify/blob/bb548d0473d4e1c9b7bbfd6602c7bf12f7a84dd2/assert/assertions.go#L1190-L1205

Reply:
Shyam - if the initial assertion requires opening a bbolt file, it is protected by the advisory flock so it appears that so further verification could be added after `assert.Panics`. 

However, this advisory lock does not block you from opening the file and perform checksum.

Reply:
sorry I stand corrected - the test indeed continues after the assert (the defer pattern was a remnant from earlier when I was calling Txn() directly without an assert) 

however, the test still got stuck after adding the assert because of the `defer betesting.Close(t, b)` at the start of the function, because it tries to write to an unbuffered channel [here](https://github.com/etcd-io/etcd/blob/3869f0e77dce84fc3da8d9385927fcdff6c87169/server/storage/backend/backend.go#L446) that is no longer read from anywhere

fixed both now, thanks!

Reply:
> I think what @serathius said is that the `assert.Panics` is capable to recover the panic. there is no need to use defer.

thx for the clarification. `defer` is still a good pattern for such case, but not a big deal though.